### PR TITLE
🧪4️⃣ test: level 4

### PR DESCRIPTION
**Pull request by florian-hoenicke**
- introduction of pypi package verification via api
- json response for more reliability

### CODE REVIEW

File: src/constants.py [hunk 1]
```diff
@@ -39,9 +38,15 @@
...
-PROBLEMATIC_PACKAGES = [
+BLACKLISTED_PACKAGES = [
     # 'Pyrender', 'Trimesh',
-    'moderngl', 'pyopengl', 'pyglet', 'pythreejs', 'panda3d', # because they need a screen,
+    'moderngl', 'pyopengl', 'pyglet', 'pythreejs', 'panda3d',  # because they need a screen,
+    'tika',  # because it needs java
+    'fastapi', 'uvicorn', 'starlette',  # because we use jina executors
 ]
 
-UNNECESSARY_PACKAGES = ['fastapi']
```
Comment (joschkabraun) [inline src/constants.py:47]: I think we should differentiate between packages which make an entire combination of packages irrelevant (as the package mustn't be used but it is functionally needed to accomplish the task) and between packages which should be skipped as they are not needed (e.g. fastapi, flask, etc.)

File: src/options/generate/templates_user.py [hunk 2]
```diff
@@ -136,13 +141,12 @@
 {code_files_wrapped}
 
 Write a single pytest case that tests the following scenario: '{test_description}'. In case the test scenario is not precise enough, test a general case without any assumptions.
-Start the test with an extensive comment about the test case. If gpt_3_5_turbo_api is used in the executor, then the test must not check the exact output of the executor as it is not deterministic. 
+Start the test with an extensive comment about the test case. If gpt_3_5_turbo is used in the executor, then the test must not check the exact output of the executor as it is not deterministic. 
 
-Use the following import to import the function:
+The test must start with the following import:
```
Comment (joschkabraun) [inline src/options/generate/templates_user.py:146]: Also encountered this issue, nice fix!

File: test/test_generator.py [hunk 3]
```diff
@@ -71,22 +68,45 @@ def test_generation_level_2(tmpdir):
...
+    GPT-3.5-turbo: ✅ (for processing the text)
+    APIs: ✅ (financial data API)
+    Databases: ❌
+    """
+    os.environ['VERBOSE'] = 'true'
+    generator = Generator(
+        f'''The input is a stock symbol (e.g., AAPL for Apple Inc.). 
+1. Fetch stock data (open, high, low, close, volume) for the past 30 days using a financial data API Yahoo Finance.
+2. Calculate the average closing price over the 30 days.
+5. Generate a brief summary of the company's stock performance over the past 30 days, including the average closing price and the company name.
```
Comment (joschkabraun) [inline test/test_generator.py:86]: Is the numbering off on purpose?

File: test/test_generator.py [hunk 4]
```diff
@@ -95,30 +115,33 @@ def test_generation_level_3(tmpdir):
...
+assert response.status_code == 200
+print('This is the text from the audio file:', response.json()['text'])
+2. Summarize the text (~50 words) while still maintaining the key facts.
+3. Create an audio file of the summarized text using a tts library.
+4. Return the the audio file as base64 encoded binary.
+Example input file: https://www.signalogic.com/melp/EngSamples/Orig/ENG_M.wav
 ''',
         str(tmpdir) + 'microservice',
-        'gpt-3.5-turbo'
+        'gpt-4'
```
Comment (joschkabraun) [inline test/test_generator.py:127]: Can turbo not solve the issue?

**Review by joschkabraun — COMMENTED** [review]
Small suggestion, otherwise LGTM!

**Comment by JoanFM** [comment]
Please, can the PR titles be more descriptive, test level X is something not understandable

File: src/constants.py (hunk 1 above)
Comment (florian-hoenicke) [reply src/constants.py:47]: good point. I updated it.

File: src/options/generate/templates_user.py (hunk 2 above)
Comment (florian-hoenicke) [reply src/options/generate/templates_user.py:146]: 👍

File: test/test_generator.py (hunk 3 above)
Comment (florian-hoenicke) [reply test/test_generator.py:86]: nice catch. I was not aware

**Comment by florian-hoenicke** [comment]
@JoanFM 
you can read more about cognitive levels here:
https://github.com/jina-ai/gptdeploy/blob/e4c680d20c2b8ece60ac2e9aed52b0162a81c8ec/test/test_generator.py#L8

File: test/test_generator.py (hunk 4 above)
Comment (florian-hoenicke) [reply test/test_generator.py:127]: yes, it was not able, but I can try again with the recent changes

Comment (joschkabraun) [reply test/test_generator.py:127]: I was just curious. Let's merge!

**Pull request merged**